0012-common-ledger-api/0012-common-ledger-api.md

```diff
@@ -0,0 +1,1201 @@
+| `connectors[].name` | [URL][]            | The unique `name` field of this connector's account. |
+| `precision`         | Integer            | How many total decimal digits of precision this ledger uses to represent currency amounts. |
+| `scale`             | Integer            | How many digits after the decimal place this ledger supports in currency amounts. |
+| `urls`              | Object             | Paths to other methods exposed by this ledger. Each field name is short name for a method and the value is the path to that method. Some fields MUST be present; see [URLs][] for details. |
+| `rounding`          | String             | _(Optional)_ The type of rounding used internally by ledger for values that exceed the reported `scale` or `precision`. If provided, the value MUST be `floor`, `ceiling`, or `nearest`. |
+| ...                 | (Various)          | _(Optional)_ Additional arbitrary values as desired. |
+
+#### Example
+
+Request:
+
+```
+GET /
+```
+
+Response:
+
+```json
+HTTP/1.1 200 OK
+
+{
+  "currency_code": "EUR",
+  "currency_symbol": "€",
+  "connectors": [{
+    "id": "https://red.ilpdemo.org/ledger/accounts/connie",
+    "name": "connie"
+  }],
+  "urls": {
+    "health": "http://usd-ledger.example/health",
+    "transfers": "http://usd-ledger.example/transfers",
+    "transfer_fulfillment": "http://usd-ledger.example/transfers/{client_id}/fulfillment",
+    "account": "http://usd-ledger.example/accounts/{name}",
+    "websocket": "wss://usd-ledger.example/websocket/"
+  },
+  "precision": 10,
+  "scale": 2
+}
+```
+
+
+#### Errors
+
+- This method does not return any errors under normal circumstances.
+
+
+### Prepare Transfer
+[Prepare Transfer]: #prepare-transfer
+
+Prepares a new transfer (conditional or unconditional) in the ledger. When a transfer becomes prepared, it executes immediately if there is no condition. If you specify `execution_condition`, the funds are held until the beneficiary [submits the matching fulfillment][Submit Fulfillment] or the `expires_at` time is reached.
+
+**Authorization:** The owner of an account MUST be able to prepare a conditional transfer that debits the account. Non-administrators MUST NOT be able to prepare a transfer that debits a different account. A ledger MAY disallow non-administrators from preparing unconditional transfers.
+
+#### Request Format
+
+POST a JSON [Transfer resource][] to the `transfers` [URL][].
+
+**Metadata URL Name:** `transfers`
+
+##### Body Parameters
+
+The message body should be a JSON [Transfer resource][].
+
+#### Response Format
+
+A successful result uses the HTTP response code **201 Created** and contains a JSON body with the [Transfer resource][] as saved.
+
+#### Example
+
+Request:
+
+```json
+POST /transfers
+Content-Type: application/json
+
+{
+  "client_id": "3a2a1d9e-8640-4d2d-b06c-84f2cd613204",
+  "ledger": "http://usd-ledger.example",
+  "debit_account": "http://usd-ledger.example/accounts/alice",
```

Comment:
The ledger is a resource identified by a URI.
The account is a resource that is relative to the ledger.

This is exactly the use case that URIs are designed for.

There is very well defined and deterministic logic for resolving absolute URIs from relative URIs given a base (the ledger) to solve for exactly the issues @clark800 raises such as changing the protocol.

If we are saying that a ledger MUST have an account endpoint URL or an account URL Template (RFC6570) then it makes sense for the accounts to be identified either by the variable that is substituted into the template or by a relative URL that is resolved against the account endpoint URL.

On the other hand if the only normative requirement is for a ledger to have a base URL then the account MUST be identified by a URL relative to the ledger URL `/accounts/alice` or an absolute URL.